test/jdk/tools/sincechecker/SinceChecker.java

```diff
@@ -0,0 +1,842 @@
+- if the given element has a @since tag in its javadoc, it is used
+- in all other cases, return the effective since value of the enclosing element
+
+
+Special Handling for preview method, as per JEP 12:
+- When an element is still marked as preview, the `@since` should be the first JDK release where the element was added.
+- If the element is no longer marked as preview, the `@since` should be the first JDK release where it was no longer preview.
+
+note on legacy preview: Until JDK 14, the preview APIs were not marked in any machine-understandable way.
+                        It was deprecated, and had a comment in the javadoc.
+                        and the use of `@PreviewFeature` only became standard in JDK 17.
+                        So the checker has an explicit knowledge of these preview elements.
+
+note: The `<unique-Element-ID>` for methods looks like
+      `method: <erased-return-descriptor> <binary-name-of-enclosing-class>.<method-name>(<ParameterDescriptor>)`.
+it is somewhat inspired from the VM Method Descriptors. But we use the erased return so that methods
+that were later generified remain the same.
+*/
+
+
+public class SinceChecker {
+    private final Map<String, Set<String>> LEGACY_PREVIEW_METHODS = new HashMap<>();
+    private final Map<String, IntroducedIn> classDictionary = new HashMap<>();
+    private final JavaCompiler tool;
+    private int errorCount = 0;
+
+    public static class IntroducedIn {
+        public String introducedPreview;
+        public String introducedStable;
+    }
+
+    public static void main(String[] args) throws Exception {
+        if (args.length == 0) {
+            throw new IllegalArgumentException("Test module not specified");
+        }
+        SinceChecker sinceCheckerTestHelper = new SinceChecker();
+        sinceCheckerTestHelper.checkModule(args[0]);
+    }
+
+    private SinceChecker() throws IOException {
+        tool = ToolProvider.getSystemJavaCompiler();
+        for (int i = 9; i <= Runtime.version().feature(); i++) {
+            JavacTask ct = (JavacTask) tool.getTask(null, null, null,
+                    List.of("--release", String.valueOf(i)),
+                    null,
+                    Collections.singletonList(SimpleJavaFileObject.forSource(URI.create("myfo:/Test.java"), "")));
+            ct.analyze();
+
+            String version = String.valueOf(i);
+            Elements elements = ct.getElements();
+            elements.getModuleElement("java.base"); // forces module graph to be instantiated
+            elements.getAllModuleElements().forEach(me ->
+                    processModuleElement(me, version, ct));
+        }
+    }
+
+    private void processModuleElement(ModuleElement moduleElement, String releaseVersion, JavacTask ct) {
+        for (ModuleElement.ExportsDirective ed : ElementFilter.exportsIn(moduleElement.getDirectives())) {
+            processElement(moduleElement, moduleElement, ct.getTypes(), releaseVersion);
```

Comment:
For consideration for tomorrow: move this one line up - to map the module id even if the module has no export directives